ddlite.py

```diff
@@ -871,47 +888,177 @@ def transform_sample_stats(Xt, t, f, Xt_abs = None):
 
-  # We initialize w at 1 for LFs & 0 for features
-  # As a default though, if no w0 provided, we initialize to all zeros
-  w = np.zeros(R) if w0 is None else w0
-  g = np.zeros(R)
-  l = np.zeros(R)
-
   # Pre-generate other matrices
   Xt = X.transpose()
   Xt_abs = abs_sparse(Xt) if sparse.issparse(Xt) else np.abs(Xt)
   
-  # Take SGD steps
-  for step in range(nSteps):
-    if step % 100 == 0 and verbose:    
-      if step % 500 == 0:
-        print "\nLearning epoch = ",
-      print "%s\t" % step,
+  # Initialize weights if no initial provided
+  w0 = np.zeros(R) if w0 is None else w0   
+  
+  # Check mixing parameter
+  if not (0 <= alpha <= 1):
+    raise ValueError("Mixing parameter must be in [0,1]")
+  
+  # Determine penalty parameters  
+  if mu_seq is not None:
+    mu_seq = np.ravel(mu_seq)
+    if not np.all(mu_seq >= 0):
+      raise ValueError("Penalty parameters must be non-negative")
+    mu_seq.sort()
+  else:
+    mu_seq = get_mu_seq(n_mu, rate, alpha, mu_min_ratio)
+
+  weights = dict()
+  # Search over penalty parameter values
+  for mu in mu_seq:
+    w = w0.copy()
+    g = np.zeros(R)
+    l = np.zeros(R)
+    # Take SGD steps
+    for step in range(maxIter):
+      if step % 100 == 0 and verbose:    
+        if step % 500 == 0:
+          print "Learning epoch = ",
+        print "%s\t" % step,
+        if (step+100) % 500 == 0:
+          print "\n",
       
+      # Get the expected LF accuracy
+      t,f = sample_data(X, w, n_samples=n_samples) if sample else exact_data(X, w)
+      p_correct, n_pred = transform_sample_stats(Xt, t, f, Xt_abs)
 
-    # Get the expected LF accuracy
-    t,f = sample_data(X, w, nSamples=nSamples) if sample else exact_data(X, w)
-    p_correct, n_pred = transform_sample_stats(Xt, t, f, Xt_abs)
+      # Get the "empirical log odds"; NB: this assumes one is correct, clamp is for sampling...
+      l = np.clip(log_odds(p_correct), -10, 10)
 
-    # Get the "empirical log odds"; NB: this assumes one is correct, clamp is for sampling...
-    l = np.clip(log_odds(p_correct), -10, 10)
+      # SGD step with normalization by the number of samples
+      g0 = (n_pred*(w - l)) / np.sum(n_pred)
 
-    # SGD step, with \ell_2 regularization, and normalization by the number of samples
-    g0 = (n_pred*(w - l)) / np.sum(n_pred) + mu*w
+      # Momentum term for faster training
+      g = 0.95*g0 + 0.05*g
 
-    # Momentum term for faster training
-    g = 0.95*g0 + 0.05*g
+      # Check for convergence
+      wn = np.linalg.norm(w, ord=2)
+      if wn < 1e-12 or np.linalg.norm(g, ord=2) / wn < tol:
+        if verbose:
+          print "SGD converged for mu={:.3f} after {} steps".format(mu, step)
+        break
+
+      # Update weights
+      w -= rate*g
+      
+      # Apply elastic net penalty
+      soft = np.abs(w) - alpha * mu
+      #          \ell_1 penalty by soft thresholding        |  \ell_2 penalty
+      w = (np.sign(w)*np.select([soft>0], [soft], default=0)) / (1+(1-alpha)*mu)
```

Comment:
Seems like it would be ~+1 lines of code to allow users to choose L1 or L2 only?